### ISSUE TITLE
Checkpointer periodically causes very high CPU load

### ISSUE THREAD

**Issue by MartinHaeusler** [cojen/Tupl#69]
Hi,
my team is using Tupl `1.3.11` on a Tomcat-based server backend with Java 8. We noticed that our server sometimes spontaneously utilizes one CPU core to almost 100% for several seconds, even though there were no requests by clients. Upon closer inspection in a profiler, we identified that the CPU load was caused by the Tupl Checkpointer thread:

![image](https://user-images.githubusercontent.com/16356998/27594020-02b56ed2-5b59-11e7-8f7e-62c1915ed7e0.png)

... which seemed strange to us, because the server was neither answering queries nor writing new data, the application logic was completely idle (please ignore the absolute numbers of the graphic above; they are very low due to the employed measurement technology). I do realize that the very purpose of a checkpointer thread is to work in the background. However, the high CPU utilization reoccurs once every couple of minutes without any modifications to the database in between. Is this a known problem, or is there any way for us to debug what exactly this is happening?

Thanks,

Martin

**Comment by broneill** [comment]
What OS are you running? What file system? Disk or SSD? When you measure CPU load, is it in system time, user time, or wait time? Java profilers don't always make a distinction, and a thread blocked on I/O might look like it's consuming CPU when it actually isn't.

If you're running Linux, trying running 'vmstat 1' for a bit to see what's going on. Also, check the size of the redo log files. If you're on Windows, simply calling 'dir' won't always report the true size, unfortunately.

**Comment by MartinHaeusler** [comment]
I am experiencing this problem under Windows 10 on an SSD. We do have Mac OS and Linux users on the team, I'll have to ask them if they see the issue as well. I actually tried running a different profiler just to see if the results match (JProfiler and JVisualVM) and I get a different picture there. I will try to use `vmstat 1` on linux too. Since the CPU load is so high, it's definitly CPU time and not wait time. I'll try to investigate further. If this is the first time you are reading about this issue, then it's likely that something else is messing us up (and fooling the profiler too).

**Comment by broneill** [comment]
If your app isn't actually writing anything, the checkpointer thread will still issue an fsync (force) call every second. The file system should recognize that nothing was actually written, and so the fsync should be a no-op. I develop on Windows and Linux, and I've never seen these no-op fsyncs cause any problems.

**Comment by broneill** [comment]
Here's a few more things to consider: Is the file or folder using NTFS encryption or compression? Do you have a virus scanner running? I'm wondering if any of these are causing the no-op fsyncs to actually trigger some other code to run.

**Comment by MartinHaeusler** [comment]
My NTFS is unencrypted and uncompressed. I do have a virus scanner, but since the windows task manager shows that the JVM is producing CPU load, I don't think that this is the issue. Honestly, I think that even tomcat itself might be at fault. JVisualVM indicated that some Tomcat I/O class was consuming high CPU, while JProfiler attributed it to Tupl's Checkpointer. Now I'm not the greatest tomcat expert, but it wouldn't surprise me if Tomcat was able to intercept the I/O calls of the webapps it is running to do some extra shenanigans. I've had an eye on this issue for quite some time now and I wasn't able to figure it out yet. Thanks for the pointers, that definitly helps.

**Comment by MartinHaeusler** [comment]
Hello again,

I've gained some insights on this issue today. I started our server and stopped the Checkpointer thread in the Eclipse debugger. The CPU usage still spiked periodically even though the Checkpointer was paused. I am 99% certain now that Tupl has nothing to do with it. The reason it popped up as a hot spot in the profiler had another cause. The profiler used a setting that excluded several packages (e.g. `com.sun`) from the measurements. The profiler then showed the "hot spots" of what remained - and that just happened to be the Tupl Checkpointer (after closer inspection, turned out that the thread did little else than sleeping, hardly consuming any CPU time at all). After turning off this filter in the profiler (and this is in line with my experiment results of pausing the checkpointer thread) it became clear that it is actually Tomcat itself that is acting up on us. That seems to be [a common issue](https://stackoverflow.com/questions/34138850/what-would-cause-tomcat-v8-to-cpu-spike-with-periodic-regularity).

Sorry for troubling you with this matter. I got fooled by the results of the profiler and jumped to conclusions. Ticket can be closed, thanks for your help.
